### PR TITLE
Handle thread exit

### PR DESCRIPTION
**Pull request by eloparco**
Implementation of what's described here: https://github.com/WebAssembly/wasi-threads/pull/17:
- a trap or `proc_exit` in the main thread should stop all the threads in the process
- a thread created with wasi_thread_spawn that calls `proc_exit` should terminate all the threads in the process
 
The sample has been modified to test combinations of trap or `proc_exit` for termination and termination forced in the main thread or a spawn thread.

Part of #1790.

### CODE REVIEW

File: core/iwasm/interpreter/wasm_runtime.c
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:2160]: Not sure why remove this line?

The wasi proc_exit native API is to terminate the process normally, but if runtime exits directly in the native API wasi_proc_exit of libc-wasi, some resources won't be destroyed, and there might be other threads running with the same module instance, normally runtime should wait for all threads of the module instance to terminate before destroy the module instance. So currently runtime just throws exception "wasi proc exit" in the native API, and clear this exception after call wasm funtion to make it exit normally.

For example, when using emsdk to build a "hello world" wasm app, normally it calls proc_exit at the end. If we don't throw "wasi proc exit" exception in the native API, then the wasm app will continue to execute the successive opcodes and "unreachable" exception will be thrown instead.

And if we don't clear "wasi_proc_exception", it will be thrown.

File: core/iwasm/common/wasm_application.c
Comment (wenyongh) [inline core/iwasm/common/wasm_application.c:113]: Should we apply the same changes to the main branch?
And should we ignore spreading the exception "wasi proc exit" since it is a normal behavior and throwing this exception is a special handling of WAMR?

Comment (eloparco) [reply core/iwasm/common/wasm_application.c:113]: This change is to propagate `proc_exit` and traps from the main thread, I can put an additional `WASM_ENABLE_LIB_WASI_THREADS != 0` at L111 to avoid changing the normal behavior when wasi threads are not involved

File: core/iwasm/interpreter/wasm_runtime.c
Comment (eloparco) [reply core/iwasm/interpreter/wasm_runtime.c:2160]: I was commenting it so that after calling `wasm_runtime_call_wasm` we can check if a `proc_exit` occurred to propagate it; since with wasi threads we have one instance per thread, do you think it would be enough to just wrap the changes with `WASM_ENABLE_LIB_WASI_THREADS`?
If not, what's your suggestion? I started contributing to this project not long ago so any clarification/suggestion is much appreciated

File: core/iwasm/common/wasm_application.c
Comment (wenyongh) [reply core/iwasm/common/wasm_application.c:113]: How about applying changes in wasm_runtime_call_wasm since here it is to call the _start function of wasm app, but developer may call other functions to start the app which may also create new threads. 

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/common/wasm_runtime_common.c#L1788-L1789

For example, before return false, add:
```C
#if WASM_ENABLE_THREAD_MGR != 0 && WASM_ENABLE_LIB_WASI_THREADS != 0
        if (wasm_runtime_get_exception(module_inst))
            wasm_cluster_spread_exception(exec_env);
#endif
```

File: core/iwasm/interpreter/wasm_runtime.c
Comment (wenyongh) [reply core/iwasm/interpreter/wasm_runtime.c:2160]: There are some places calling clear_wasi_proc_exit_exception, I think if you need to disable the clear exception action, had better just modify the clear_wasi_proc_exit_exception itself, in both wasm_runtime.c and aot_runtime:

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#L2037

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/aot/aot_runtime.c#L903

File: core/iwasm/common/wasm_application.c
Comment (eloparco) [reply core/iwasm/common/wasm_application.c]: yes that makes more sense, thanks

**Review by wenyongh — APPROVED** [review]
LGTM

File: core/iwasm/interpreter/wasm_runtime.c
Comment (loganek) [inline core/iwasm/interpreter/wasm_runtime.c:2036]: should we then capture the exception and clear it for each module instances?

Comment (eloparco) [reply core/iwasm/interpreter/wasm_runtime.c:2036]: I think I need to modify `wasm_runtime_call_wasm()` to return `true` in case the exception encountered is `proc_exit`

File: samples/wasi-threads/wasm-apps/thread_termination.c
Comment (eloparco) [inline samples/wasi-threads/wasm-apps/thread_termination.c:34]: @wenyongh btw I notice something while testing: if I replace this for loop with a `__builtin_wasm_memory_atomic_wait32(0, 0, -1);` or even `while(1);` the thread doesn't get interrupted even if the exception is propagated (in the first case it gets stuck here https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/common/posix/posix_thread.c#L192). Is that expected? I'll investigate to understand what's happening

File: core/iwasm/interpreter/wasm_runtime.c
Comment (yamt) [reply core/iwasm/interpreter/wasm_runtime.c:2036]: my expectation on `proc_exit` is:

* wasm_application_execute_main etc succeeds without an exception
* exit code is propagated here https://github.com/bytecodealliance/wasm-micro-runtime/blob/219713bfbc5196ee7c91fe78d18ddceed069c179/product-mini/platforms/posix/main.c#L669

File: samples/wasi-threads/wasm-apps/thread_termination.c
Comment (yamt) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: it's what i expect to happen with the current implementation.

to terminate busy looping threads, something like `wasm_cluster_thread_send_signal` would work.
i guess we eventually need to implement something (eg. send a signal on posix-like platforms) to terminate the blocking threads.

Comment (wenyongh) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: > @wenyongh btw I notice something while testing: if I replace this for loop with a `__builtin_wasm_memory_atomic_wait32(0, 0, -1);` or even `while(1);` the thread doesn't get interrupted even if the exception is propagated (in the first case it gets stuck here https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/common/posix/posix_thread.c#L192). Is that expected? I'll investigate to understand what's happening

Normally runtime should terminate all threads when exception was thrown, I remember it is the expected behavior of Java VM. Seems it is an issue of the main branch, I try to fix it with the patch below for interpreter. And another issue of main branch may be that the main thread doesn't spread exception to other thread when exception was thrown, like the change of this PR in wasm_runtime_call_wasm. I will discuss with @xujuntwt95329 for how to fix them.
```diff
diff --git a/core/iwasm/interpreter/wasm_interp_fast.c b/core/iwasm/interpreter/wasm_interp_fast.c
index 3109b0c8..36863bf2 100644
--- a/core/iwasm/interpreter/wasm_interp_fast.c
+++ b/core/iwasm/interpreter/wasm_interp_fast.c
@@ -1054,6 +1054,8 @@ wasm_interp_call_func_import(WASMModuleInstance *module_inst,
 #if WASM_ENABLE_THREAD_MGR != 0
 #define CHECK_SUSPEND_FLAGS()                           \
     do {                                                \
+        if (wasm_get_exception(module))                 \
+            goto got_exception;                         \
         if (exec_env->suspend_flags.flags != 0) {       \
             if (exec_env->suspend_flags.flags & 0x01) { \
```

Comment (wenyongh) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: @eloparco I discussed the issues with @xujuntwt95329, he agreed that they are issues and will submit patch to fix them on main branch. Could you wait until we fix them and merge branch `main` into `dev/wasi_threads`, so as to sync up the code and avoid conflicts in the future?

File: core/iwasm/interpreter/wasm_runtime.c
Comment (loganek) [reply core/iwasm/interpreter/wasm_runtime.c:2036]: > I get an exception so the runtime doesn't return 0 at the end 

> it seems like the exit code gets already propagated properly

I'm a bit confused, so does it return the exit code being passed to `proc_exit()`?

Comment (eloparco) [reply core/iwasm/interpreter/wasm_runtime.c:2036]: yes, the return code is set when `proc_exit` is called https://github.com/bytecodealliance/wasm-micro-runtime/blob/903f5212f194c2709b32631bd32936fdcfed27af/core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c#L989 and then retrieved here https://github.com/bytecodealliance/wasm-micro-runtime/blob/219713bfbc5196ee7c91fe78d18ddceed069c179/product-mini/platforms/posix/main.c#L669
I tested it by calling `__wasi_proc_exit(exit_code)` with different exit codes and I see the `iwasm` executable returning with the proper exit codes

File: samples/wasi-threads/wasm-apps/thread_termination.c
Comment (eloparco) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: as proposed by @yamt, I tried adding a `wasm_cluster_send_signal_all(cluster, WAMR_SIG_TERM);` after spreading the exception https://github.com/bytecodealliance/wasm-micro-runtime/blob/903f5212f194c2709b32631bd32936fdcfed27af/core/iwasm/libraries/thread-mgr/thread_manager.c#L995 and it solves in case of a `while(1);` but not for other cases like `sleep(TIMEOUT_SECONDS);` or `__builtin_wasm_memory_atomic_wait32(0, 0, -1);`

Comment (yamt) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: > but not for other cases like `sleep(TIMEOUT_SECONDS);` or `__builtin_wasm_memory_atomic_wait32(0, 0, -1);`

i guess something more complex is necessary for blocking threads.
like use `pthread_kill` and make them handle `EINTR`.

for another runtime i made all i/o non-blocking to handle thread termination requests. but i don't think it's an appropriate approach for WAMR.

Comment (xujuntwt95329) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: @eloparco Yes, `wasm_cluster_send_signal_all` can only tell the interpreter to exit the execution loop, so `while(1)` can be terminated, but for `sleep` and `__builtin_wasm_memory_atomic_wait32`, they are not executed in interpreter, they have already fall into OS kernel, so the `wasm_cluster_send_signal_all` will not work for this situation.

I agree with @yamt that we may need some more complex mechanism to terminate the threads when they are blocked in kernel.

Comment (loganek) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: Is that also a problem in lib_pthread? I wonder if we should fix it in the `main` branch as well.

Comment (eloparco) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: > Is that also a problem in lib_pthread?

yes; it applies in general to threads running "blocking" instructions like `sleep()` or `__builtin_wasm_memory_atomic_wait32()`

> I wonder if we should fix it in the main branch as well.

I think that's what @wenyongh was proposing a few comments above

File: core/iwasm/aot/aot_runtime.c
Comment (xujuntwt95329) [inline core/iwasm/aot/aot_runtime.c:907]: Seems `lib-pthread` also need to spread the proc exit exception, I'll modify this line to `#if WASM_ENABLE_LIBC_WASI != 0 && WASM_ENABLE_THREAD_MGR == 0` in main branch

File: samples/wasi-threads/wasm-apps/thread_termination.c
Comment (wenyongh) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: @loganek @eloparco Jun's PR #1889 was merged and I tried to create a PR to merge branch main into dev/wasi_threads, but unfortunately there are conflicts in below files:
```
        .github/workflows/compilation_on_android_ubuntu.yml
        core/iwasm/libraries/thread-mgr/thread_manager.c
        core/iwasm/libraries/thread-mgr/thread_manager.h
```
I can resolve these conflicts, and the diff file of the PR is as below:
[merge_main_into_wasi_threads.zip](https://github.com/bytecodealliance/wasm-micro-runtime/files/10462938/merge_main_into_wasi_threads.zip)

Could you help check whether the patch works for you? If yes, I will push it manually under the command line. Thanks.

Comment (loganek) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: Hi @wenyongh , I reviewed the patch and it is mostly correct, except one thing. In function `thread_manager_start_routine()` you moved `free_aux_stack` below `wasm_runtime_deinstantiate_internal`; `free_aux_stack` must be executed first because it uses module instance. 

Also, one minor thing about `wasm_cluster_add_exec_env` - the `ret` variable is no longer needed so we could remove it; but that's not a big deal.

You can either update your merge commit or feel free to merge it as it is, and I'll open a PR to address the problem.

Comment (wenyongh) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: @loganek Thanks for the review, I changed the wasm_runtime_deinstantiate_internal in thread_manager_start_routine. But for the wasm_cluster_add_exec_env, not sure why `ret` can be removed.
I pushed the PR, see: https://github.com/bytecodealliance/wasm-micro-runtime/commit/c7141894fb7b5ef5d39f13748382c9138ee945de

Comment (loganek) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: > But for the wasm_cluster_add_exec_env, not sure why ret can be removed.

It can be replaced with `return` statements (we no longer have a lock to unlock). As I said, it's not a big deal though. The rest looks ok for me.

Comment (loganek) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: > I agree with @yamt that we may need some more complex mechanism to terminate the threads when they are blocked in kernel.

@xujuntwt95329 / @wenyongh does anybody on your side actively working on that? If not, we'll take on that work.

Comment (wenyongh) [reply samples/wasi-threads/wasm-apps/thread_termination.c:34]: @loganek not yet

**Comment by eloparco** [comment]
We can probably merge this PR and I'll open a separate issue to track the possible race condition that I documented above (and I've never been able to reproduce again).

**Pull request merged**
